### PR TITLE
Add `LLVM` support for no wrap arithmetic operations

### LINKED ISSUE #181: Add `LLVM` support for `nsw` arithmetic operations
@jclark for this I am planning of adding fallowing changes to the public API

1. Add new `public type BinaryNSWArithmeticOp "add"|"mul"|"sub"` (This is because not all of the current `BinaryArithmeticOp`s have a `nsw` version
2. Add new function to builder as `public function binaryNSWInt(BinaryNSWArithmeticOp op, Value lhs, Value rhs, string? name=()) returns Value` (Similar to `binaryInt` in current implementation but since not all `BinaryIntOp`s have a `nsw` version)

2 Will represent fallowing C-API functions:

- [`LLVMBuildNSWAdd`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#ga901f44572f6ef9f67e5f7bb496806280)
- [`LLVMBuildNSWSub`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#ga0ababa21464265587c66120354f857f0)
- [`LLVMBuildNSWMul`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#ga620c57b22a134350df2cce5e676d2b88)

I am assuming we don't need the equivalent to [`LLVMBuildNSWNeg`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#gaac84a51432d611b64b3570fc36067a48) at the moment.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose

- Refactor binary integer operations
- Add support to no wrap arithmetic operations

Resolves #181

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/nback/build.bal [hunk 1]
```diff
@@ -523,13 +523,13 @@ function buildArithmeticBinary(llvm:Builder builder, Scaffold scaffold, bir:IntA
         builder.positionAtEnd(continueBlock);
         continueBlock = scaffold.addBasicBlock();
         llvm:BasicBlock overflowBlock = scaffold.addBasicBlock();
-        builder.condBr(builder.binaryInt("and",
+        builder.condBr(builder.iBitwise("and",
                                          builder.iCmp("eq", lhs, llvm:constInt(LLVM_INT, int:MIN_VALUE)),
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: Please fix the indentation also: the `"and"` and `builder.` should align.

File: compiler/modules/nback/build.bal [hunk 2]
```diff
@@ -758,7 +758,7 @@ function buildConvertRepr(llvm:Builder builder, Scaffold scaffold, Repr sourceRe
 
 function buildTaggedBoolean(llvm:Builder builder, llvm:Value value) returns llvm:Value {
     return builder.getElementPtr(llvm:constNull(LLVM_TAGGED_PTR),
-                                     [builder.binaryInt("or",
+                                     [builder.iBitwise("or",
                                                         builder.zExt(value, LLVM_INT),
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: Fix indentation

File: compiler/modules/print.llvm/common.bal [hunk 3]
```diff
@@ -18,7 +18,7 @@ public type IntegralType IntType|PointerType;
 
 // Corresponds to LLVMPointerType function
 public function pointerType(Type ty, int addressSpace = 0) returns PointerType {
-    return { pointsTo: ty, addressSpace };
+    return {pointsTo: ty, addressSpace};
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Please use consistent spacing style which has space after `{` and before `}` when inline.

File: compiler/modules/print.llvm/common.bal [hunk 4]
```diff
@@ -28,7 +28,7 @@ public type ArrayType readonly & record {|
 |};
 
 public function arrayType(Type ty, int elementCount) returns ArrayType {
-    return { elementType: ty, elementCount: elementCount };
+    return {elementType: ty, elementCount: elementCount};
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Fix spacing style

File: compiler/modules/print.llvm/common.bal [hunk 5]
```diff
@@ -37,7 +37,7 @@ public type StructType readonly & record {
 };
 
 public function structType(Type[] elementTypes) returns StructType {
-    return { elementTypes: elementTypes.cloneReadOnly() };
+    return {elementTypes: elementTypes.cloneReadOnly()};
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Fix spacing style

File: compiler/modules/print.llvm/common.bal [hunk 6]
```diff
@@ -55,16 +55,21 @@ public type FunctionType readonly & record {|
...
-public type BinaryArithmeticOp "add"|"mul"|"sub"|"sdiv"|"srem";
-public type BinaryBitwiseOp "xor"|"or"|"and";
+
+public type IntArithmeticOp "add"|"sub"|"mul";
+
+public type IntArithmeticSignedOp "sdiv"|"srem";
+
+public type IntBitwiseOp "xor"|"or"|"and";
+
+public type IntOp IntArithmeticOp|IntArithmeticSignedOp|IntBitwiseOp;
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Do we use `IntOp`?

File: llvm_jni/modules/jni.llvm/Builder.bal [hunk 7]
```diff
@@ -76,7 +76,39 @@ public distinct class Builder {
         }
     }
 
-    public function binaryInt(BinaryIntOp op, Value lhs, Value rhs, string? name = ()) returns Value {
+    public function iArithmeticNoWrap(IntArithmeticOp op, Value lhs, Value rhs, string? name=()) returns Value {
+        string regName = self.extractName(name);
+        match op {
+            "add" => {
+                Value val = new (jLLVMBuildNSWAdd(self.LLVMBuilder, lhs.LLVMValueRef, rhs.LLVMValueRef, java:fromString(regName)));
```
Comment (jclark) [inline llvm_jni/modules/jni.llvm/Builder.bal:83]: Does LLVMBuildNSW* add both `nsw` and `nuw`?

File: llvm_jni/modules/jni.llvm/tests/testOutputs/arithmetic_no_wrap.ll [hunk 8]
```diff
@@ -0,0 +1,6 @@
+define i64 @foo(i64 %0, i64 %1) {
+  %3 = add nsw i64 %0, %1
```
Comment (jclark) [inline llvm_jni/modules/jni.llvm/tests/testOutputs/arithmetic_no_wrap.ll:2]: Looks like this is adding just `nsw`. I think we want `nsw nuw`.  Can you get that from the C API?

**Review by jclark — CHANGES_REQUESTED** [review]
Main problem I see is that the JNI version is generating just `nsw` whereas the print version is generating `nuw nsw`.

Apart from that just a few nits.

File: llvm_jni/modules/jni.llvm/tests/testOutputs/arithmetic_no_wrap.ll (hunk 8 above)
Comment (heshanpadmasiri) [reply llvm_jni/modules/jni.llvm/tests/testOutputs/arithmetic_no_wrap.ll:2]: @jclark unfortunately C-API has only has [`LLVMBuildNSWAdd`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#ga901f44572f6ef9f67e5f7bb496806280) and [`LLVMBuildNUWAdd`](https://llvm.org/doxygen/group__LLVMCCoreInstructionBuilder.html#gac46472d120e1a0d83222509e4e418ac2) nothing that can do both. (Only the C++ API at the moment can do both with [`CreateAdd`](https://llvm.org/doxygen/classllvm_1_1IRBuilderBase.html#a928603739e0e70713566011d44052a4f)

File: compiler/modules/nback/build.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/build.bal]: Fixed

File: compiler/modules/nback/build.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/build.bal]: Fixed

File: compiler/modules/print.llvm/common.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: Fixed

File: compiler/modules/print.llvm/common.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: Fixed

File: compiler/modules/print.llvm/common.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: Fixed

File: compiler/modules/print.llvm/common.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: It is use internally withing the builder only removed `public`.

File: llvm_jni/modules/jni.llvm/Builder.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply llvm_jni/modules/jni.llvm/Builder.bal:83]: No unfortunately with the C-API as far as I can work out we can do only `nsw` or `nuw` not both

File: compiler/modules/print.llvm/llvm.bal [hunk 9]
```diff
@@ -344,9 +344,31 @@ public class Builder {
...
 
-    // binary operation with int operands and (same) int result
-    // Corresponds to LLVMBuild{Add,Mul,Sub,SDiv,SRem}
-    public function binaryInt(BinaryIntOp op, Value lhs, Value rhs, string? name=()) returns Value {
+    // Corresponds to LLVMBuildNSW{Add,Mul,Sub}
+    public function iArithmeticNoWrap(IntArithmeticOp op, Value lhs, Value rhs, string? name=()) returns Value {
+        BasicBlock bb = self.bb();
+        string reg = bb.func.genReg();
+        IntType ty = sameIntType(lhs, rhs);
+        bb.addInsn(reg, "=", op, "nuw", "nsw", ty, lhs.operand, ",", rhs.operand);
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: I think we should do just "nsw" here to keep the two llvm modules consistent.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/print.llvm/llvm.bal (hunk 9 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: Fixed

**Pull request merged**